### DISCUSSION TITLE
Generic Linux

### DISCUSSION THREAD

**Discussion by JonBoyleCoding** [midchildan/dotfiles#108] [Q&A]
I've noticed in a couple of places you have a `genericLinux` specified in the config targets.

https://github.com/midchildan/dotfiles/blob/fae87a3ef327c23031d8081333678f9472e4c0ed/nix/home/profiles/development.nix#L5

After a brief search I couldn't find this being set anywhere - is this something that you manually set on a machine basis? I could only find references to this and could not find where this was being created or being set.

I just wanted to check as I've been doing this manually - but if there was a way to automatically detect this it would be amazing.

**Comment by midchildan** [comment]
You're right, it's something that needs to be set on a per-machine basis. This is because of an intentional design choice by Nix. It actively tries to remove so-called impurities during evaluation of configuration files. In other words, the values of Home Manager options are determined solely by Nix code written or imported in the configuration and nothing else. Automatic feature detection during deployment is therefore not possible here. But in exchange, deployments are more reproducible, predictable, and inspectable as a result.

If you want automatic detection, it would need to be done prior to deployment. The results can then be dumped to a file, checked into git, and imported by Nix code.

If detection really needs to be done during deployment, dropping flakes may be an option too. Pre-flake Nix allows some impurities with builtin functions such as `getEnv` and `readFile`. But those loopholes are closed with flakes. This dotfiles also provides a top level `default.nix` file, so it should be compatible with non-flakes usage too.
